pkg/kubelet/metrics/collectors/emptydir_metrics.go

```diff
@@ -0,0 +1,142 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collectors
+
+import (
+	"context"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/component-base/metrics"
+	"k8s.io/klog/v2"
+
+	kubeletmetrics "k8s.io/kubernetes/pkg/kubelet/metrics"
+	serverstats "k8s.io/kubernetes/pkg/kubelet/server/stats"
+)
+
+var (
+	emptyDirUsedBytesDesc = metrics.NewDesc(
+		metrics.BuildFQName(
+			"",
+			kubeletmetrics.KubeletSubsystem,
+			kubeletmetrics.EmptyDirUsedBytesKey,
+		),
+		"Bytes used by the emptyDir volume.",
+		[]string{
+			"volume_name",
+			"namespace",
+			"pod",
+		},
+		nil,
+		metrics.ALPHA,
+		"",
+	)
+	emptyDirSizeLimitBytesDesc = metrics.NewDesc(
+		metrics.BuildFQName(
+			"",
+			kubeletmetrics.KubeletSubsystem,
+			kubeletmetrics.EmptyDirSizeLimitBytesKey,
+		),
+		"Size limit of the emptyDir volume in bytes, if set.",
+		[]string{
+			"volume_name",
+			"namespace",
+			"pod",
+		},
+		nil,
+		metrics.ALPHA,
+		"",
+	)
+)
+
+type emptyDirMetricsCollector struct {
+	metrics.BaseStableCollector
+
+	statsProvider serverstats.Provider
+}
+
+// Check if emptyDirMetricsCollector implements necessary interface
+var _ metrics.StableCollector = &emptyDirMetricsCollector{}
+
+// NewEmptyDirMetricsCollector implements the metrics.StableCollector interface and
+// exposes metrics about pod's emptyDir.
+func NewEmptyDirMetricsCollector(statsProvider serverstats.Provider) metrics.StableCollector {
+	return &emptyDirMetricsCollector{statsProvider: statsProvider}
+}
+
+// DescribeWithStability implements the metrics.StableCollector interface.
+func (c *emptyDirMetricsCollector) DescribeWithStability(ch chan<- *metrics.Desc) {
+	ch <- emptyDirUsedBytesDesc
+	ch <- emptyDirSizeLimitBytesDesc
+}
+
+// CollectWithStability implements the metrics.StableCollector interface.
+func (c *emptyDirMetricsCollector) CollectWithStability(ch chan<- metrics.Metric) {
+	podStats, err := c.statsProvider.ListPodStats(context.Background())
+	if err != nil {
+		klog.ErrorS(err, "Failed to get pod stats")
+		return
+	}
+
+	for _, podStat := range podStats {
+		podName := podStat.PodRef.Name
+		podNamespace := podStat.PodRef.Namespace
+
+		if podStat.VolumeStats == nil {
+			klog.V(5).InfoS("Pod has no volume stats", "pod", podName, "namespace", podNamespace)
+			continue
+		}
+
+		pod, found := c.statsProvider.GetPodByName(podNamespace, podName)
+		if !found {
+			klog.V(5).InfoS("Couldn't get pod", "pod", podName, "namespace", podNamespace)
+			continue
+		}
+
+		podVolumes := make(map[string]v1.Volume, len(pod.Spec.Volumes))
+		for _, volume := range pod.Spec.Volumes {
+			podVolumes[volume.Name] = volume
+		}
+
+		for _, volumeStat := range podStat.VolumeStats {
+			if volume, found := podVolumes[volumeStat.Name]; found {
+				if volume.EmptyDir != nil && volume.EmptyDir.Medium == v1.StorageMediumDefault {
+					if volumeStat.UsedBytes != nil {
```

Comment:
Should we expose sth if `nil`?
`NaN` or sth. Or log sth?

Reply:
I am not too familiar with what that would mean in practice, but wouldn't "0" be suited here?

Reply:
Maybe it could be `nil` temporarily (e.g., while the Pod is being created). I expect it will eventually not be nil. We can probably leave it as is, similar to how it's handled in `log_metrics.go` https://github.com/kubernetes/kubernetes/blob/bbd83d86444d7b325a51f4daa0d65163b795b70e/pkg/kubelet/metrics/collectors/log_metrics.go#L74 (we can always revisit them later)